### ISSUE TITLE
feat: give autocmd callback arg a `win` field

### ISSUE THREAD

**Issue by tmillr** [neovim/neovim#23581]
### Problem

Autocmd callbacks don't receive a window or win id. For example, this makes it very difficult, if not impossible, to correctly detect which window an option was set in when using an `OptionSet` autocmd.

I have a couple different use-cases atm. One is to refrain from setting (window) options if the opt in question was last set from a modeline. AFAIK, the only way to detect this is to use an `OptionSet` autocmd, which will tell you how the option was set (e.g. `modeline`), but not which window it was set on (am I to assume that it will simply always be the current window within/during the callback?). Yes, it already gives a buffer id as `buf`, but a buffer may have multiple windows, each with different window options, so a window id would be helpful here.

### Expected behavior

Similar to how we already have a `buf` field, give autocmd callback argument tables a `win` field denoting the relevant window. Optionally, do this for all autocmd events (or all where it would make sense), but at the very least, do it for `OptionSet`.

**Comment by zeertzjq** [comment]
> am I to assume that it will simply always be the current window within/during the callback?

Yes.

**Comment by tmillr** [comment]
> > am I to assume that it will simply always be the current window within/during the callback?
> 
> Yes.

Oh interesting. Thank you, that's good to know

**Comment by ii14** [comment]
@nyngwang It's true only for some events. In case of `WinClosed` the window ID is in the `"match"` field, you just have to use `tonumber` to turn it into a number.

**Comment by nyngwang** [comment]
@ii14 But in the real case I provided above the current window upon the `WinClosed` event is `1005` while the `match` is `1663`, which is apparently different. So I still disagree with the conclusion:

> "[...] assume that it will simply always be the current window within/during the callback."

If this is true then why do they differ?

**Comment by ii14** [comment]
This was specifically in the context of OptionSet event, and you're talking about WinClosed event, so that's different. Events are inconsistent in this and it isn't great, but that's just how it works right now and it's hard to change anything without breaking compatibility. For OptionSet you just get the current window and for WinClosed you get the window ID from the "match" field.

**Comment by tmillr** [comment]
> This was specifically in the context of OptionSet event, and you're talking about WinClosed event, so that's different. Events are inconsistent in this and it isn't great, but that's just how it works right now and it's hard to change anything without breaking compatibility. For OptionSet you just get the current window and for WinClosed you get the window ID from the "match" field.

Yeah, that's why I was kinda surprised that `OptionSet` is actually called within the target window. If you read the docs, it seems to stress over and over how you cannot/shouldn't rely on the current buffer or window ever being the target buf/win. But w/e, at least it works for what I needed, even if it is awkward and confusing/inconsistent.

**Comment by zeertzjq** [comment]
Oh, autocommands can change the current window, so this may make sense.

**Comment by justinmk** [comment]
Also may be useful to have `ev.tab`. https://github.com/neovim/neovim/pull/38537#issuecomment-4321799702

**Comment by justinmk** [comment]
tracked in:

- https://github.com/neovim/neovim/issues/25844
